src/services/utilities.ts

```diff
@@ -927,4 +933,101 @@ namespace ts {
+        const result: VisibleModuleInfo[] = [];
+        findPackageJsons(scriptPath).forEach((packageJson) => {
+            const package = tryReadingPackageJson(packageJson);
+            if (!package) {
+                return;
+            }
+
+            const nodeModulesDir = combinePaths(getDirectoryPath(packageJson), "node_modules");
+            const foundModuleNames: string[] = [];
+
+            if (package.dependencies) {
+                addPotentialPackageNames(package.dependencies, modulePrefix, foundModuleNames);
+            }
+            if (package.devDependencies) {
+                addPotentialPackageNames(package.devDependencies, modulePrefix, foundModuleNames);
+            }
+
+            forEach(foundModuleNames, (moduleName) => {
+                const moduleDir = combinePaths(nodeModulesDir, moduleName);
+                result.push({
+                    moduleName,
+                    moduleDir,
+                    canBeImported: moduleCanBeImported(moduleDir)
+                });
+            });
+        });
+
+        return result;
+
+        function findPackageJsons(currentDir: string): string[] {
+            const paths: string[] = [];
+            let currentConfigPath: string;
+            while (true) {
+                currentConfigPath = findConfigFile(currentDir, (f) => host.fileExists(f), "package.json");
+                if (currentConfigPath) {
+                    paths.push(currentConfigPath);
+
+                    currentDir = getDirectoryPath(currentConfigPath);
+                    const parent = getDirectoryPath(currentDir);
+                    if (currentDir === parent) {
+                        break;
+                    }
+                    currentDir = parent;
+                }
+                else {
+                    break;
+                }
+            }
+
+            return paths;
+        }
+
+        function tryReadingPackageJson(filePath: string) {
+            try {
+                const fileText = host.readFile(filePath);
+                return JSON.parse(fileText);
+            }
+            catch (e) {
+                return undefined;
+            }
+        }
+
+        function addPotentialPackageNames(dependencies: any, prefix: string, result: string[]) {
+            for (const dep in dependencies) {
+                if (dependencies.hasOwnProperty(dep) && (!prefix || startsWith(dep, prefix))) {
+                    result.push(dep);
+                }
+            }
+        }
+
+        /*
+         * A module can be imported by name alone if one of the following is true:
+         *     It defines the "typings" property in its package.json
+         *     The module has a "main" export and an index.d.ts file
+         *     The module has an index.ts
+         */
+        function moduleCanBeImported(modulePath: string): boolean {

```

Comment:
Some modules can't be required by their name alone, but instead define a bunch of submodules. For example, the "angular2" package cannot be required just the package name:
```Typescript
import * as a from "angular2/core";     // Works fine
import * as b from "angular2";          // Error
```
In that case, it's more helpful to return a completion with a trailing slash (e.g. "angular2/") or just give completions for all of the nested names. If this check is too expensive, I'm fine with removing it since that really is a minor thing.